Fix invalid byte sequence in UTF-8 for sendgrid

staffler-xyz opened the pull request:
There is an error when the filename for sendgrid contains an invalid character, for example "test\xc3\x28.pdf".

staffler-xyz committed:
Update adapter.rb

sdhull commented:
Can you add a spec for this please?

staffler-xyz committed:
added spec

staffler-xyz committed:
added forced encoding to text

staffler-xyz committed:
fixed typo

sdhull commented on lib/griddler/sendgrid/adapter.rb:
```diff
@@ -21,7 +21,7 @@ def normalize_params
             score: params[:spam_score],
           }
-
+          text: params[:text]&.to_s&.force_encoding(Encoding::UTF_8)
```
Does this line have a spec I'm not seeing? Why is `.force_encoding(Encoding::UTF_8)` good enough here but line 82 below requires `filename.force_encoding('ISO-8859-1').encode('UTF-8')`?

staffler-xyz committed:
convert to utf-8 automatically

staffler-xyz commented:
Hi, I just updated the PR. I think it's the best to convert all strings to UTF-8 before sending the data to rails. What I changed:
- charset is automatically set to utf-8
- convert all string values to utf-8 (check validity first, replace invalid characters if needed)

What do you think?

staffler-xyz committed:
removed unused line

staffler-xyz committed:
only replace certain parameters

sdhull commented on lib/griddler/sendgrid/adapter.rb:
```diff
@@ -49,7 +68,8 @@ def bcc_from_envelope
       def charsets
         return {} unless params[:charsets].present?
-        JSON.parse(params[:charsets]).symbolize_keys
+        charsets = JSON.parse(params[:charsets]).symbolize_keys
+        charsets.transform_values!{ "UTF-8" }
```
looking at this, it seems potentially destructive, or possibly not i18n-friendly ... 🤔 is there some valid case where someone might want _not_ utf8 charsets?

At this point should we introduce a configuration option, so all this behavior is opt-in?

sdhull reviewed (commented):
Apologies for the late review, you know how work goes sometimes 😓 

Some questions for you but overall I like the direction and suspect it will be great for ~80% of folks as-is. I just want to make sure we're not causing issues for the other ~20%

stefannibrasil commented:
Closing this as this library has been deprecated: https://github.com/thoughtbot/griddler-sendgrid/pull/41
